spanner/ot_metrics.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+import (
+	"context"
+	"log"
+	"strconv"
+	"strings"
+	"sync"
+
+	"cloud.google.com/go/spanner/internal"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"google.golang.org/grpc/metadata"
+)
+
+// OtInstrumentationScope is the instrumentation name that will be associated with the emitted telemetry.
+const OtInstrumentationScope = "cloud.google.com/go"
+const metricsPrefix = "spanner/"
+
+var (
+	attributeKeyClientID   = attribute.Key("client_id")
+	attributeKeyDatabase   = attribute.Key("database")
+	attributeKeyInstance   = attribute.Key("instance_id")
+	attributeKeyLibVersion = attribute.Key("library_version")
+	attributeKeyType       = attribute.Key("type")
+	attributeKeyMethod     = attribute.Key("grpc_client_method")
+
+	attributeNumInUseSessions = attributeKeyType.String("num_in_use_sessions")
+	attributeNumSessions      = attributeKeyType.String("num_sessions")
+	// openTelemetryMetricsEnabled is used to track if OpenTelemetry Metrics need to be recorded
+	openTelemetryMetricsEnabled = false
```

Comment:
These are all global variables. What is our strategy to prevent both errors and confusion if someone tries to enable/disable open telemetry multiple times in their application. For example, what happens if someone:
1. Enables Open Telemetry.
2. Creates a Spanner client
3. Disables Open Telemetry
4. Creates another Spanner client

Is the above allowed? If so, what will be the behavior of the two Spanner clients?

Reply:
These are global un-exported variables which are not available for public access. We expose only one public method `EnableOpenTelemetryMetrics()` which is called by customers once at the beginning of their application to use OpenTelemetry instrumentation for metrics. There is no method to disable OpenTelemetry. So there is no way to enable and disable OpenTelemetry multiple times.

**The example mentioned above is not possible**.

Possible Scenarios
1. Scenario 1:
       -  Enables Open Telemetry
       - Creates Spanner client 1
       - Creates Spanner client 2
      
       Observed Behavior - OpenTelemetry metrics get recorded for both the spanner clients.
2. Scenario 2:
       - Creates Spanner client 1
       -  Enables Open Telemetry
       - Creates Spanner client 2
      
       Observed Behavior - OpenTelemetry metrics get recorded only for spanner client 2.

Reply:
Ah, thanks for explaining. That looks good to me.